pi.py

```diff
@@ -1 +1,578 @@
+        jmin=int(min([jmin,j]))
+    
+        #
+        #   *** Calculate Parcel quantities BELOW lifted condensation level   ***
+        #
+        if (P[j] >= PLCL):
+            # Parcel temperature at this pressure
+            TG=TP*(P[j]/PP)**(RD/CPD)
+            # Parcel Mixing ratio
+            RG=RP
+            # Parcel and Environmental Density Temperatures at this pressure (E94, EQN. 4.3.1 and 6.3.7)
+            TLVR=TG*(1.+RG/EPS)/(1.+RG)
+            TVENV=T[j]*(1.+R[j]/EPS)/(1+R[j])
+            # Bouyancy of the parcel in the environment (Proxy of E94, EQN. 6.1.5)
+            TVRDIF[j,]=TLVR-TVENV
+            
+        #
+        #   *** Calculate Parcel quantities ABOVE lifted condensation level   ***
+        # 
+        else:
+            
+            # Initial default values before loop
+            TGNEW=T[j]
+            TJC=T[j]-273.15
+            ES=6.112*np.exp(17.67*TJC/(243.5+TJC))
+            RG=EPS*ES/(P[j]-ES)
+            
+            #
+            #   ***  Iteratively calculate lifted parcel temperature and mixing   ***
+            #   ***                ratio for reversible ascent                    ***
+            #
+            
+            # set loop counter and initial condition
+            NC=0
+            TG=0
+
+            # loop until loop converges or bails out
+            while ((np.abs(TGNEW-TG)) > 0.001):
+            
+                # Parcel temperature and mixing ratio during this iteration
+                TG=TGNEW
+                TC=TG-273.15
+                ENEW=6.112*np.exp(17.67*TC/(243.5+TC))
+                RG=EPS*ENEW/(P[j]-ENEW)
+                
+                # increase iteration count in the loop
+                NC=NC+1
+                
+                #
+                #   ***  Calculate estimates of the rates of change of the entropy    ***
+                #   ***           with temperature at constant pressure               ***
+                #
+
+                ALV=ALV0+CPVMCL*TC
+                SL=(CPD+RP*CL+ALV*ALV*RG/(RV*TG*TG))/TG
+                EM=RG*P[j]/(EPS+RG)
+                # (last term vanishes with saturation, i.e. RH=1)
+                SG=(CPD+RP*CL)*np.log(TG)-RD*np.log(P[j]-EM)+ALV*RG/TG
+                # convergence speed (AP, step in entropy fraction) varies as a function of 
+                # number of iterations
+                if (NC < 3):
+                    # converge slowly with a smaller step
+                    AP=0.3
+                else:
+                    # speed the process with a larger step when nearing convergence
+                    AP=1.0
+                # find the new temperature in the iteration
+                TGNEW=TG+AP*(S-SG)/SL
+                
+                #
+                #   ***   If the routine does not converge, set IFLAG=2 and bail out   ***
+                #
+                if (NC > 500) or (ENEW > (P[j]-1)):
```

Comment:
When would the routine not converge?

Reply:
I wrote about this in the documentation. It very rarely happens (Kerry notes that it NEVER happened in the analyses Bister and Emanuel 2002, and it doesn't happen for my example either, see the doc). It probably relates to some oddity in the input profile? This would be a good thing to investigate at some point, but its so rare that I haven't wanted to spend much time on it.

Reply:
We could systematically look at all of the NaNs you are finding in your results and dig into what those input profiles look like, tracking through the code?

Reply:
I'll have to dig up the analyses where I found this and may not get to it later, but I'll keep it in the back of my mind. Thanks.

Reply:
Please do! Its a mystery I haven't had occasion to solve, but would be useful to document.

Reply:
OK... I looked a bit into this since I was getting crazy values of PI for some real-time forecasts. Turns out I had an indexing error in my code,  which was offset by another error in that I was inputting the wrong units of MSLP into the PI code. After fixing those two errors, I don't get the NaNs in the PI outputs anymore.